### DISCUSSION TITLE
Cannot archive build on Xcode 13 beta 3

### DISCUSSION THREAD

**Discussion by d-date** [pointfreeco/combine-schedulers#32] [General] [answered]
**Describe the bug**

When running archive build on Xcode 13 beta 3, the build fail since it depends on Combine.framework built by different compiler version.

> Failed to build module 'Combine'; this SDK is not supported by the compiler (the SDK is built with 'Apple Swift version 5.5 (swiftlang-1300.0.24.14 clang-1300.0.25.10)', while this compiler is 'Apple Swift version 5.5 (swiftlang-1300.0.24.13 clang-1300.0.25.10)'). Please select a toolchain which matches the SDK.

**Screenshots**
<img width="864" alt="Screen Shot 2021-07-15 at 13 36 50" src="https://user-images.githubusercontent.com/20222809/125729582-34a9ce13-f47f-4850-9716-bf8f4bf6905f.png">

**Environment**
 - Xcode 13 beta 3
 - Swift 5.5

**Comment by stephencelis** [comment]
Hi @d-date! I'm able to successfully build a project that depends on CombineSchedulers in beta 3. The error message you posted makes it sound like you have a custom toolchain selected in Xcode. Can you check the "Xcode > Toolchains" menu and verify that "Xcode 13.0" is selected?

The only other thing I can think of is maybe you have some stale build products getting in the way. "Product > Clean Build Folder" could help, or you could try fully [deleting the Derived Data](https://programmingwithswift.com/delete-derived-data-xcode/) folder.

**Comment by d-date** [comment]
@stephencelis  
Did you succeed in the **Archive**, right?
 I'm able to build and run, but archive has failed.

As you say, I checked that Xcode 13 was selected as the toolchain. I also tried a clean build, but didn't change.

<img width="789" alt="Screen Shot 2021-07-15 at 22 56 49" src="https://user-images.githubusercontent.com/20222809/125800252-dffc0d9b-b167-4d67-a5d6-784d0a3ac404.png">

<img width="783" alt="Screen Shot 2021-07-15 at 23 01 03" src="https://user-images.githubusercontent.com/20222809/125800867-177991c7-1d98-46bd-8e14-7550fbabc4c7.png">

**Comment by kamcma** [comment]
I am also seeing this error. Although to start I am suspecting beta 3 or maybe something about my workspace configuration more than combine-schedulers per se.

I don't think I have any alternative toolchains installed. (The toolchain items don't even appear in the Xcode menu or Preferences > Components.) Deleted derived data. M1 Mac, if it matters.

**Comment by kamcma** [comment]
Seems to be the same issue:

https://developer.apple.com/forums/thread/682285?answerId=682006022#682006022

**Comment by mbrandonw** [comment]
😢 oof, good to know. 

I'm going to covert this to a discussion since it's not really a CombineSchedulers issue and so that we can continue sharing information about it.

**Comment by tgrapperon** [comment]
I also had the issue while trying to profile a TCA iOS15 only app with Xcode 13b3. I've added platform availability in a local copy of `Package.swift` and it solved the issue:
```
// swift-tools-version:5.1

import PackageDescription

let package = Package(
  name: "combine-schedulers",
  platforms: [
    .iOS(.v13),
    .macOS(.v10_15),
    .tvOS(.v13),
    .watchOS(.v6),
  ],
  products: [...
```

**Comment by schneiderandre** [comment]
Unfortunately, the issue is still present in Xcode 13 beta 4, but Apple is already aware of the issue.

Since SPM does currently not allow to replace a specific sub-dependency with a fork, I for now forked:
- https://github.com/pointfreeco/combine-schedulers, changed the package platforms information to 

```swift
.platforms: [
    .iOS(.v12),
    .macOS(.v10_14),
    .tvOS(.v12),
    .watchOS(.v5)
  ]
```
and replaced the sub dependency in the package in a fork of:
- http://github.com/pointfreeco/swift-composable-architecture

While this is not optimal, it is at least a functioning workaround for now, locally and on the CI.
I am unsure if we should directly have a branch with this temporary fix in both main repositories to make it easier for others or wait and assume that the next beta will address the issue.

**Reply by maximkrouk** [reply]
Btw version on `main` branch should not require specifying os versions in a Package.swift file, so probably releasing `0.5.1` may fix the issue (but I'm not sure)

**Reply by stephencelis** [reply]
Good point! We just released it: https://github.com/pointfreeco/combine-schedulers/releases/tag/0.5.1

**Reply by petrpavlik** [reply]
`0.5.1` does not seem to fix the issue for me using Xcode 13 beta 4.

**Reply by schneiderandre** [reply]
Hey @petrpavlik, true. This release just allows older platform versions to work as well. Once the compiler error is fixed, this is a good addition, but until then we need to limit the platforms (e.g. via a fork, or separate branch) like mentioned above.

**Reply by mycroftcanner** [reply]
![Screenshot 2021-08-24 at 22 37 26](https://user-images.githubusercontent.com/55385694/130693471-871404f9-f886-4ba1-823c-19834b7f8f38.png)

Xcode 13 beta 5

Is there anyway to fix this besides forking all the packages? Profiling and archiving is broken.

**Reply by maximkrouk** [reply]
@mycroftcanner, do you use the package as a part of another package or as a part of an xcodeproj?

**Reply by mycroftcanner** [reply]
I have a big Packages.swift similar to the one in is-words.

**Reply by mycroftcanner** [reply]
Specifying 0.5.1 doesn't help

**Reply by mbrandonw** [reply]
> Is there anyway to fix this besides forking all the packages? Profiling and archiving is broken.

@mycroftcanner Do you know of a fix for this issue? What would you do in the forked packages? We're definitely interested in fixing this issue, we're just not sure what the solution is.

**Reply by tgrapperon** [reply]
@mbrandonw Specifying `platforms` in `Package.swift` seems to fix the issue. Maybe TCA can temporarily rely on a branch of `combine-schedulers` with this addition until it's resolved by Apple? 
It seems to affect packages that are importing Combine, like [Alamofire](https://github.com/Alamofire/Alamofire/issues/3481), [GRDB](https://github.com/groue/GRDB.swift/issues/1033) or [Realm](https://github.com/realm/realm-cocoa/issues/7368#issuecomment-895296330).

**Comment by tgrapperon** [comment]
I've asked why the "platforms" field had to be removed, but then read the original PR, understood the reason, and deleted my question.
Unfortunately, @maximkrouk had already responded. Here is the response:

Nope, but it can be a part of the app that does support iOS13- and uses it just for iOS13+ features and you won't be able to use it at all if the platform is defined by the package (at least if an app or framework use spm-based modularization)

Anyway it seems like an SPM bug and I have something strange here too https://github.com/MakeupStudio/cocoa-aliases/blob/main/Sources/CocoaAliases/CocoaAlisaes.swift (when building in release mode the compiler cannot find UIViewRepresentable)

**Reply by maximkrouk** [reply]
I couldn't find your comment, so I deleted the reply too 😅

**Reply by stephencelis** [reply]
Sounds like we should revert the platform removal to work around this bug. Does that make sense to you @maximkrouk?

**Reply by maximkrouk** [reply]
Yep
1. Seems like it is the only workaround for now
2. The bug will probably affect me too when migrating to XCode 13 and packages that I currently make (`combine-schedulers -> combine-extensions -> combine-cocoa -> swift-extensions -> composable-architecture-extensions`) are private (no one depends on them yet, even I) and probably they will suffer from the bug too, so I'll have to add platforms to use them
3. It will be pretty easy to remove platforms when the bug is fixed

**Comment by stephencelis** [comment, marked as answer]
We just published [0.5.2](https://github.com/pointfreeco/combine-schedulers/releases/tag/0.5.2), which will hopefully resolve the issue.

**Reply by nmccann** [reply]
@stephencelis 0.5.2 seems to have resolved the previous issues mentioned in this discussion, however I'm still encountering archiving issues on Xcode 13 Beta 5 - specifically: `SourcePackages/checkouts/combine-schedulers/Sources/CombineSchedulers/FailingScheduler.swift:77:5: cannot find 'XCTFail' in scope`

In my Google searching for the issue, I also found this possibly related solution: https://github.com/groue/CombineExpectations/pull/15/files

**Reply by stephencelis** [reply]
@nmccann You may need to bump XCTestDynamicOverlay, too. What version are you on?

**Reply by nmccann** [reply]
@stephencelis I think you're correct - looking at my Package.resolved, I'm on 0.1.0, not the latest (0.2.1) - I'll ensure those are updated and try it again. Thank you for the quick response.

**Reply by nmccann** [reply]
Resolving the latest versions resolved my issue, thanks again!
